src/memo/memo.ml

```diff
@@ -190,19 +198,30 @@ module M = struct
          [Init] should be treated exactly the same as [Running*] with a stale
          value of [Run.t]. *)
       | Init
-      | Running_sync : Run.t -> ('a, 'b, 'a -> 'b) t
-      | Running_async : Run.t * 'b Fiber.Ivar.t -> ('a, 'b, 'a -> 'b Fiber.t) t
+      | Running_sync : Run.t * Deps_so_far.t -> ('a, 'b, 'a -> 'b) t
+      | Running_async :
+          Run.t * Deps_so_far.t * 'b Fiber.Ivar.t
+          -> ('a, 'b, 'a -> 'b Fiber.t) t
       | Failed of Run.t * Exn_with_backtrace.t
       | Done of 'b Cached_value.t
   end =
     State
 
-  and Dep_node : sig
+  and Dep_node_without_state : sig
     type ('a, 'b, 'f) t =
       { spec : ('a, 'b, 'f) Spec.t
       ; input : 'a
       ; id : Id.t
       ; dag_node : Dag.node
```

Comment:
As we discussed, we should try to move this field into `State.t` to make `Dep_node` squarely correspond to signals and `Dag.node` correspond to samplings. This might have a side benefit of dropping the reference to `dag_node` when the state is `Done`.